test/ecto/schema_test.exs

```diff
@@ -516,6 +633,42 @@ defmodule Ecto.SchemaTest do
     assert inspect(comment) == "#Ecto.Association.NotLoaded<association :comment is not loaded>"
   end
 
+  test "belongs_to association via query" do
+    struct =
+      %Ecto.Association.BelongsTo{field: :high_scoring_comment, owner: AssocSchema, cardinality: :one,
+       related: Comment, owner_key: :comment_id, related_key: :id,
+       queryable: Comment.high_score(), on_replace: :raise, defaults: []}
+
+    assert AssocSchema.__schema__(:association, :high_scoring_comment) == struct
+    assert AssocSchema.__changeset__.high_scoring_comment == {:assoc, struct}
+
+    comment = (%AssocSchema{}).high_scoring_comment
+    assert %Ecto.Association.NotLoaded{} = comment
+    assert inspect(comment) == "#Ecto.Association.NotLoaded<association :high_scoring_comment is not loaded>"
+  end
+
+  test "many_to_many association via query" do
+    struct = %Ecto.Association.ManyToMany{
+      field: :awesome_editors,
+      owner: AssocSchema,
+      cardinality: :many,
+      on_delete: :nothing,
+      related: User,
+      owner_key: :id,
+      queryable: User.awesome(),
+      on_replace: :delete,
+      join_keys: [assoc_schema_id: :id, user_id: :id],
+      join_through: AssocEditors.active()
+    }
+
+    assert AssocSchema.__schema__(:association, :awesome_editors) == struct
+    assert AssocSchema.__changeset__.awesome_editors == {:assoc, struct}
+
+    awesome_editors = (%AssocSchema{}).awesome_editors
+    assert %Ecto.Association.NotLoaded{__cardinality__: :many} = awesome_editors
+    assert inspect(awesome_editors) == "#Ecto.Association.NotLoaded<association :awesome_editors is not loaded>"
+  end
```

Comment:
Those tests are testing implementation details and not any behaviour and I avoid those kind of tests generally. For example, we may rename a field, everything will continue to work, but we would have failing tests. I suggest to remove the tests added to this file.